art/defences/trainer/certified_trainer_pytorch.py

```diff
@@ -0,0 +1,372 @@
+
+                    if all(certification_results):
+                        samples_certified += 1
+
+                    if (i + 1) % batch_size == 0 and i > 0:
+                        break
+
+                certified_loss /= batch_size
+                # Concrete PGD loss
+                i_batch = np.copy(
+                    x_preprocessed[ind[m * self.pgd_params["batch_size"] : (m + 1) * self.pgd_params["batch_size"]]]
+                ).astype("float32")
+                o_batch = y_preprocessed[
+                    ind[m * self.pgd_params["batch_size"] : (m + 1) * self.pgd_params["batch_size"]]
+                ]
+
+                # Perform prediction
+                self.set_forward_mode("concrete")
+                self.attack = ProjectedGradientDescent(
+                    estimator=self._classifier,
+                    eps=self.pgd_params["eps"],
+                    eps_step=self.pgd_params["eps_step"],
+                    max_iter=self.pgd_params["max_iter"],
+                    num_random_init=self.pgd_params["num_random_init"],
+                )
+                i_batch = self.attack.generate(i_batch, y=o_batch)
+                self._classifier.model.zero_grad()
+                model_outputs = self._classifier.model.forward(i_batch)
+                acc = self._classifier.get_accuracy(model_outputs, o_batch)
+
+                # Form the loss function
+                pgd_loss = self._classifier.concrete_loss(
+                    model_outputs, torch.from_numpy(o_batch).to(self._classifier.device)
+                )  # pylint: disable=W0212
+
+                pbar.set_description(
+                    f"Bound {bound:.2f}: Loss {pgd_loss:.2f} Cert Loss {certified_loss:.2f} "
+                    f"Acc {acc:.2f} Cert Acc {samples_certified / batch_size:.2f}"
+                )
+
+                loss = certified_loss * self.loss_weighting + pgd_loss * (1 - self.loss_weighting)
+                # Do training
+                if self._classifier._use_amp:  # pragma: no cover # pylint: disable=W0212
+                    from apex import amp  # pylint: disable=E0611
+
+                    with amp.scale_loss(loss, self._classifier._optimizer) as scaled_loss:  # pylint: disable=W0212
+                        scaled_loss.backward()
+
+                else:
+                    loss.backward()
+
+                self._classifier._optimizer.step()  # pylint: disable=W0212
+
+            if scheduler is not None:
+                scheduler.step()
+
+    def predict(self, x: np.ndarray, **kwargs) -> Union["torch.Tensor", Tuple["torch.Tensor", "torch.Tensor"]]:
+        """
+        Perform prediction using the adversarially trained classifier.
+
+        :param x: Input samples.
+        :param kwargs: Other parameters to be passed on to the `predict` function of the classifier.
+        :return: Predictions for test set.
+        """
+        return self._classifier.model.forward(x, **kwargs)
```

Comment:
I suggest to use the `classifier.predict` method.

Reply:
predict and predict_zonotopes now included